Envoy cidr hosts

jrajahalme opened the pull request:
Support CIDRs in NPHDS in addition to plain host addresses. This allows CIDR-based HTTP policy enforcement for both ingress and egress.

Cilium daemon is not yet updated to send down CIDRs in NPHDS, but integration testing is added to exercise the new functionality. Integration test is changed to not depend on configuration files to reduce clutter in the repo.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/3763)
<!-- Reviewable:end -->

rlenglet commented on envoy/cilium_host_map.cc:
```diff
@@ -11,12 +11,124 @@
+    auto it = maps.begin();
+    for (; it != maps.end(); it++) {
+      if (it->first > plen) {
+	ENVOY_LOG(debug, "Skipping map for prefix length {} while looking for {}", it->first, plen);
```
The debug logs in this method are going to be VERY verbose. How about using `trace` instead?

jrajahalme replied:
Good point. I added them for tracing purposes anyway!

rlenglet commented on envoy/cilium_host_map.cc:
```diff
@@ -11,12 +11,124 @@
+      const char *slash = strchr(addr, '/');
+      if (slash != nullptr) {
+	const char *pstr = slash + 1;
+	// Must start with a non-zero digit
```
Why not just call `atoi(pstr)` or `strtoul(pstr, NULL, 10)` and then check that it's within the acceptable range for the IP version?

jrajahalme replied:
They accept all kinds of stuff like whitespaces, plus and minus signs, and garbage after the prefix length. I did not want to be that lax in parsing.

rlenglet replied:
OK.

rlenglet commented on envoy/cilium_host_map.h:
```diff
@@ -1,19 +1,49 @@
+  return addr & ~hton((I(1) << (PLEN_MAX - plen)) - 1);
+};
+
+enum ID : uint64_t { UNKNOWN = 0, HOST = 1, WORLD = 2, CLUSTER = 3 };
```
Why do you need to have those defined in the filter? I only see `ID:WORLD` used.
You're supposed to receive the mapping for "host" through NPHDS. I think it should be the case for "cluster" too.
OK with defining constants for "unknown" and "world", but I don't think you should define an enum.

jrajahalme replied:
enum is the method for defining constants, and I guess I defined the unused pieces to give context to the one we actually use. Alternatively we could require the nphds server insert a default ("/0") CIDR for the world and just return 0 (== UNKNOWN) if there is no match at all.

jrajahalme replied:
Alternatively we could make the "not found" case return value ID configurable in the filter config?

jrajahalme replied:
IMO it would be cleanest to require the default CIDR to be configured in the host map itself and simply return 0 for the "not found" case.

jrajahalme replied:
OK, I removed the unused values (HOST, CLUSTER) from the enum for now. The default return for the "not found" case is still ID::WORLD (2), but now CIDRs with 0 prefix lengths are supported. So when Cilium daemon is changed to pass in "0.0.0.0/0" and "::/0" CIDRs we can remove the WORLD definition and simply return 0 when an address cannot be mapped to any ID.

rlenglet commented on envoy/cilium_host_map.cc:
```diff
@@ -11,12 +11,124 @@
+	ENVOY_LOG(debug, "Found existing map for prefix length {}", plen);
+	return it->second;
+      }
+      // Current pair has has smaller prefix, insert before it to maintain order
```
has has -> has

jrajahalme replied:
Fixed, thanks!

rlenglet reviewed (approved):
LGTM! I only have relatively minor comments.

jrajahalme committed:
envoy: Fix integration test setting of original dst address.

Original dst address was as with '0.0.0.0' (or '::' for IPv6), which
worked for ingress tests. Egress tests need the actual address, so set
the original_dst_address to the canonical loopback address instead.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
envoy: Use network byte-order addresses in host map.

This is more efficient and is a step into making the keys maskable for
CIDR lookup.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
envoy: Support CIDRs in NPHDS.

Allow optional prefix length (e.g., "/28") at the end of host addresses.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
envoy: Egress intergation tests.

Verify that egress proxy policy enforcement works for the same test
cases as ingress. Remote identity is resolved from the host map, for
which a matching CIDR entry is configured.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
envoy: Add host map to cilium integration test

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>
